nixos/modules/services/desktops/pipewire.nix

```
@@ -65,38 +92,81 @@ in {
-        @func getenv
-        vars [ PCM ]
-        default "plug:pipewire"
-        playback_mode "-1"
-        capture_mode "-1"
-      }
-    '';
+    environment.etc."alsa/conf.d/49-pipewire-modules.conf" = mkIf cfg.alsa.enable {
+      text = ''
+        pcm_type.pipewire {
+          libs.native = ${cfg.package.lib}/lib/alsa-lib/libasound_module_pcm_pipewire.so ;
+          ${optionalString enable32BitAlsaPlugins
+            "libs.32Bit = ${pkgs.pkgsi686Linux.pipewire.lib}/lib/alsa-lib/libasound_module_pcm_pipewire.so ;"}
+        }
+        ctl_type.pipewire {
+          libs.native = ${cfg.package.lib}/lib/alsa-lib/libasound_module_ctl_pipewire.so ;
+          ${optionalString enable32BitAlsaPlugins
+            "libs.32Bit = ${pkgs.pkgsi686Linux.pipewire.lib}/lib/alsa-lib/libasound_module_ctl_pipewire.so ;"}
+        }
+      '';
+    };
     environment.etc."alsa/conf.d/50-pipewire.conf" = mkIf cfg.alsa.enable {
-      source = "${pkgs.pipewire}/share/alsa/alsa.conf.d/50-pipewire.conf";
+      source = "${cfg.package}/share/alsa/alsa.conf.d/50-pipewire.conf";
+    };
+    environment.etc."alsa/conf.d/99-pipewire-default.conf" = mkIf cfg.alsa.enable {
+      source = "${cfg.package}/share/alsa/alsa.conf.d/99-pipewire-default.conf";
     };
     environment.sessionVariables.LD_LIBRARY_PATH =
       lib.optional cfg.jack.enable "/run/current-system/sw/lib/pipewire";
+
+    environment.etc."pipewire/pipewire.conf" = {
+      # Adapted from src/daemon/pipewire.conf.in
+      text = ''
+        set-prop link.max-buffers 16 # version < 3 clients can't handle more
+
+        add-spa-lib audio.convert* audioconvert/libspa-audioconvert
+        add-spa-lib api.alsa.* alsa/libspa-alsa
+        add-spa-lib api.v4l2.* v4l2/libspa-v4l2
+        add-spa-lib api.libcamera.* libcamera/libspa-libcamera
+        add-spa-lib api.bluez5.* bluez5/libspa-bluez5
+        add-spa-lib api.vulkan.* vulkan/libspa-vulkan
+        add-spa-lib api.jack.* jack/libspa-jack
+        add-spa-lib support.* support/libspa-support
+
+        load-module libpipewire-module-rtkit # rt.prio=20 rt.time.soft=200000 rt.time.hard=200000
+        load-module libpipewire-module-protocol-native
+        load-module libpipewire-module-profiler
+        load-module libpipewire-module-metadata
+        load-module libpipewire-module-spa-device-factory
+        load-module libpipewire-module-spa-node-factory
+        load-module libpipewire-module-client-node
+        load-module libpipewire-module-client-device
+        load-module libpipewire-module-portal
+        load-module libpipewire-module-access
+        load-module libpipewire-module-adapter
+        load-module libpipewire-module-link-factory
+        load-module libpipewire-module-session-manager
+
+        create-object spa-node-factory factory.name=support.node.driver node.name=Dummy priority.driver=8000
+
+        exec ${cfg.package}/bin/pipewire-media-session ${lib.concatStringsSep " " cfg.mediaSessionExtraArguments}
```

Maybe it would be better to make this whole line configurable?  Wireplumber can replace `pipewire-media-session`: https://pipewire.pages.freedesktop.org/wireplumber/daemon/running.html?gi-language=c
```suggestion
        exec wireplumber
```

---

Sounds good, but then the whole compat setup thing should be changed a bit.
For pipewire.conf I guess `exec ${cfg.sessionManager} ${lib.concatStringsSep " " cfg.sessionManagerArguments}` would do, but I don't know what the best way to handle `/etc/pipewire/media-session.d`?

---

Maybe pipewire-media-session should be split into its own package? Then these options and the  `/etc/pipewire/media-session.d/` stuff could perhaps be done there? And switching the sessionManager package to wireplumber with whatever config setup that wants would maybe Just Work...

---

Ok, I've added the `sessionManager` and `sessionManagerArguments` options.  But yes, wireplumber support won't be as easy as setting `sessionManager`; it needs its own configuration file.

---

~~Hmm should this be handled by something like `extraEtcFiles`?~~ Or, on second thought, maybe this should be handled entirely on the module level, like `services.xserver.displayManager.*`? Something like `services.pipewire.sessionManager.{pipewireSessionManager, wireplumber}`. Considering that displayManager is a pretty visible precedent maybe that's the best solution already for consistency reasons.

---

Just want to mention that the `exec` line could be dropped and session manager launched via systemd-user instead.

---

> Something like services.pipewire.sessionManager.{pipewireMediaSession, wireplumber}

I think we should wait with designing such an elaborate system until wireplumber/pipewire stabilizes a bit.  For now, the current options are enough to 1) enable optional features in pipewire-media-session, and 2) change/disable the session manager if you want to run wireplumber manually.

Anything beyond this should probably wait at least until we have a wireplumber package.

> Just want to mention that the `exec` line could be dropped and session manager launched via systemd-user instead.

That seems like a neat idea.  I'm surprised that upstream doesn't do that by default.